### DISCUSSION TITLE
Providing server-side props without serializing and sending to the client

### DISCUSSION THREAD

**Discussion by rmunch** [vercel/next.js#27555] [Ideas]
### Describe the feature you'd like to request

I've run into a couple of use cases where I need to pass a value from `getServerSideProps` to the page component only during SSR, _without_ having that value serialized and sent to the client.

* In the first use case, I want to provide sensitive request headers to the page component. While it's possible to read request headers in `getServerSideProps`, because my app is cached at the CDN, the fact that server-side props are serialized and sent to the client means that those sensitive values will be cached and served to visitors who should not see those values.
* In the second case, I want to initialize a value for each request, and make that same instance available to both `getServerSideProps` and the page component. The value is something that can't be serialized, so as expected, returning the value in `getServerSideProps` leads to the error `getServerSideProps cannot be serialized as JSON`. My desired solution is for the browser to freshly initialize its own new instance rather than attempting to serialize and hydrate the value.

### Describe the solution you'd like

Some solutions I've thought of would be:

* Provide a way to selectively tell NextJS which props to serialize and send to the client. This would have some design impacts - for example, this might require a pattern around initializing those missing props in the browser.
* As discussed at length in https://github.com/vercel/next.js/issues/11993, provide a way to tell NextJS how to serialize values. This would allow me to serialize these values to `null` or similar.
* For the sensitive request header issue, if I could somehow access request headers directly in my page component (and not just via `getServerSideProps`), that would also work. This seems like a less generalized solution.

### Describe alternatives you've considered

For now, I've worked around the unserializable value issue by providing a shared singleton via module imports. This works when the singleton can be shared for all requests, though my requirements are now changing where I'll need to create a new instance for each request.

~~I don't have a workaround for passing sensitive request headers to page components, but I did find a potential workaround at https://github.com/vercel/next.js/issues/7393. Since this solution uses `getInitialProps` in the main App, it looks like this will disable SSG, so is less ideal.~~ Looks like this solution still ends up serializing props and sending to the client.

**Comment by programmatix** [comment]
To add another use-case - I simply want to use Next, for some pages, to do a standard dynamic server page.  E.g. just render some server-side content, using React JSX, without hydration.  Yes, I could roll this myself using ReactDOMServer, but a) Next provides some very useful functionality that would take a while to emulate, such as Fast Refresh and the automatic routing, and b) for some other pages I will want the full SSR hydration.
